app/notification_push.go

```diff
@@ -270,20 +306,49 @@ func (a *App) StopPushNotificationsHubWorkers() {
 	}
 }
 
-func (a *App) sendToPushProxy(msg model.PushNotification, session *model.Session) {
+func (a *App) sendToPushProxy(msg model.PushNotification, session *model.Session) error {
 	msg.ServerId = a.DiagnosticId()
 
-	request, err := http.NewRequest("POST", strings.TrimRight(*a.Config().EmailSettings.PushNotificationServer, "/")+model.API_URL_SUFFIX_V1+"/send_push", strings.NewReader(msg.ToJson()))
+	var updatedMsg *model.PushNotification
+
+	if pluginsEnvironment := a.GetPluginsEnvironment(); pluginsEnvironment != nil {
+		pluginContext := a.PluginContext()
+		pluginsEnvironment.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
+			updatedMsg = hooks.PushNotificationWillBeSent(pluginContext, &msg)
+			return true
+		}, plugin.PushNotificationWillBeSentId)
+	}
+	mlog.Debug(fmt.Sprintf("Push notification %s will be sent", updatedMsg.Id))
+	mlog.Debug(fmt.Sprintf("Push notification %s will be sent to device %s", updatedMsg.Id, updatedMsg.DeviceId))
+
+	if updatedMsg == nil {
+		return nil
+	}
+
+	request, err := http.NewRequest("POST", strings.TrimRight(*a.Config().EmailSettings.PushNotificationServer, "/")+model.API_URL_SUFFIX_V1+"/send_push", strings.NewReader(updatedMsg.ToJson()))
 	if err != nil {
-		mlog.Error(fmt.Sprintf("Error sending to push proxy: UserId=%v SessionId=%v message=%v",
-			session.UserId, session.Id, err.Error()), mlog.String("user_id", session.UserId))
-		return
+		if pluginsEnvironment := a.GetPluginsEnvironment(); pluginsEnvironment != nil {
```

Comment:
1/5 for this failure (failed to compose an HTTP request) invoking the plugin is an overkill.

Reply:
Is not about the failure composing the HTTP request, is about the failure trying to send the notification. If I try to send the notification and for any reason is not sent, I need to know from the plugins side (If we want to have the possibility of monitor entirely the push notification process).

Reply:
These should practically never happen anyway. So as long as the boilerplate code is compacted I don't mind.